# Pandas

`pandas` is an open-source module for Python that contains data structures and data manipulation tools designed to make data cleaning and analysis fast and convenient in Python.

It adopts significant parts of NumPy's idiomatic style of array-based computing, especially array-based functions and a preference for data processing without for loops.

While `pandas` adopts many coding idioms from `NumPy`, the biggest difference is that `pandas` is designed for working with tabular or heterogeneous data. `NumPy`, by contrast, is best suited for working with homogeneously typed numerical array data

## Installation

Run the following in the terminal:
`pip install pandas`

If running inside Jupyter/Colab notebook, run:
`!pip install pandas`

## Getting Started

Two central concepts in `pandas` are: `Series` and `DataFrame`.

- A `Series` is like a vector or a column containing some values of the same datatype
- A `DataFrame` is a table consisting of one or more `Series`. Each series can have its own datatype

## `Series`
- A Series is a one-dimensional array-like object containing a sequence of values (of similar types to NumPy types) of the same type
- Every `Series` has an associated array of data labels, called its `index`

In [1]:
import pandas as pd

In [2]:
obj = pd.Series([4, 7, -5, 3])
obj

,0
0,4
1,7
2,-5
3,3


In [3]:
obj[0:2]    # we can perform slicing just like in numpy

,0
0,4
1,7


Since we did not specify an index for the data, a default one consisting of the integers 0 through N-1 (where N is the length of the data) is created

We can also create a Series with an index identifying each data point with a label:

In [4]:
obj2 = pd.Series([4, 7, -5, 3], index=["d", "b", "a", "c"])
obj2

,0
d,4
b,7
a,-5
c,3


In [5]:
obj2["b"]

np.int64(7)

## `DataFrame`
A `DataFrame` represents a rectangular table of data and contains an ordered, named collection of columns, each of which can be a different value type (numeric, string, Boolean, etc.).

There are many ways to construct a `DataFrame` from scratch, though one of the most common is from a dictionary (`dict`) of equal-length `list`s or `NumPy` arrays

In [6]:
data = {"state": ["Ohio", "Ohio", "Ohio", "Nevada", "Nevada", "Nevada"],
        "year": [2000, 2001, 2002, 2001, 2002, 2003],
        "pop": [1.5, 1.7, 3.6, 2.4, 2.9, 3.2]}
frame = pd.DataFrame(data)
frame

,state,year,pop
0,Ohio,2000,1.5
1,Ohio,2001,1.7
2,Ohio,2002,3.6
3,Nevada,2001,2.4
4,Nevada,2002,2.9
5,Nevada,2003,3.2


The resulting `DataFrame` has its `index` assigned automatically, as with `Series`, and the columns are placed according to the order of the keys in data (which depends on their insertion order in the dictionary)

## Reading and Writing Data using `pandas`

| Function       | Description                                                                                              |
|----------------|----------------------------------------------------------------------------------------------------------|
| read_csv       | Load delimited data from a file, URL, or file-like object; use comma as default delimiter                |
| read_clipboard | Variation of read_csv that reads data from the clipboard; useful for converting tables from web pages    |
| read_excel     | Read tabular data from an Excel XLS or XLSX file                                                         |
| read_json      | Read data from a JSON (JavaScript Object Notation) string representation, file, URL, or file-like object |
| read_sas       | Read a SAS dataset stored in one of the SAS system's custom storage formats                              |
| read_spss      | Read a data file created by SPSS                                                                         |
| read_stata     | Read a dataset from Stata file format                                                                    |


## Loading the data

We shall now load the IMDb Top 1000 Movies Dataset (https://www.kaggle.com/datasets/moazeldsokyx/imdb-top-10000-movies-dataset). It contains rating, runtime, genre, metascore, plot, directors, stars, votes, gross earnings etc of 9999 movies up to June 2023.

In [7]:
movies_df = pd.read_csv("https://github.com/ami-null/python-course/raw/refs/heads/main/data/Top_10000_Movies_IMDb.csv")

After loading the data, we should view the basic informations regarding the columns:

In [8]:
movies_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9999 entries, 0 to 9998
Data columns (total 12 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   ID          9999 non-null   int64  
 1   Movie Name  9999 non-null   object 
 2   Rating      9999 non-null   float64
 3   Runtime     9999 non-null   object 
 4   Genre       9999 non-null   object 
 5   Metascore   7992 non-null   float64
 6   Plot        9999 non-null   object 
 7   Directors   9999 non-null   object 
 8   Stars       9999 non-null   object 
 9   Votes       9999 non-null   int64  
 10  Gross       9999 non-null   int64  
 11  Link        9999 non-null   object 
dtypes: float64(2), int64(3), object(7)
memory usage: 937.5+ KB


In [9]:
movies_df.shape

(9999, 12)

We see that the data has 12 columns, 9999 rows, and only the `Metascore` column has missing values. It also shows the data type of each column. The columns with `Dtype` as `object` can be strings, lists etc.

We should also print out a few rows of the dataset to keep as a visual reference. We accomplish this with `.head()` method:

In [10]:
movies_df.head()

,ID,Movie Name,Rating,Runtime,Genre,Metascore,Plot,Directors,Stars,Votes,Gross,Link
0,1,The Shawshank Redemption,9.3,142 min,Drama,82.0,"Over the course of several years, two convicts...","['Frank Darabont', 'Tim Robbins', 'Morgan Free...","['Tim Robbins', 'Morgan Freeman', 'Bob Gunton'...",2752419,28341469,https://www.imdb.com/title/tt0111161/
1,2,The Godfather,9.2,175 min,"Crime, Drama",100.0,"Don Vito Corleone, head of a mafia family, dec...","['Francis Ford Coppola', 'Marlon Brando', 'Al ...","['Marlon Brando', 'Al Pacino', 'James Caan', '...",1914751,134966411,https://www.imdb.com/title/tt0068646/
2,3,Ramayana: The Legend of Prince Rama,9.2,135 min,"Animation, Action, Adventure",NaN,An anime adaptation of the Hindu epic the Rama...,"['Ram Mohan', 'Yûgô Sakô', 'Koichi Saski', 'Ar...","['Yûgô Sakô', 'Koichi Saski', 'Arun Govil', 'N...",10108,10108,https://www.imdb.com/title/tt0259534/
3,4,The Chaos Class,9.2,87 min,"Comedy, Drama",NaN,"Lazy, uneducated students share a very close b...","['Ertem Egilmez', 'Kemal Sunal', 'Münir Özkul'...","['Kemal Sunal', 'Münir Özkul', 'Halit Akçatepe...",41785,41785,https://www.imdb.com/title/tt0252487/
4,5,Daman,9.1,121 min,"Adventure, Drama",NaN,"The film is set in 2015. Sid, is a young docto...","['Lenka Debiprasad', 'Vishal Mourya', 'Karan K...","['Vishal Mourya', 'Karan Kandhapan', 'Babushan...",13303,13303,https://www.imdb.com/title/tt17592606/


`.head()` outputs the first five rows of a DataFrame by default, but we could also pass a number as well: `movies_df.head(10)` would output the first ten rows, for example.

Further, in the above output, we see that the `Metascore` column has some `NaN` values. `NaN` means "not a number" and indicates missing values in the case of numeric columns. In columns with other data types, `None` or `<NA>` might be used to indicated missing values.

In [11]:
# we don't want spaces inside column names
movies_df = movies_df.rename(columns = {"Movie Name" : "title"})

To see the last five rows, we can use `.tail()` method.

`.tail()` also accepts a number, and in this case we printing the bottom three rows:

In [12]:
movies_df.tail(3)

,ID,title,Rating,Runtime,Genre,Metascore,Plot,Directors,Stars,Votes,Gross,Link
9996,9997,Paranormal Activity 4,4.6,88 min,"Horror, Mystery",40.0,It has been five years since the disappearance...,"['Henry Joost', 'Ariel Schulman', 'Stephen Dun...","['Ariel Schulman', 'Stephen Dunham', 'Katie Fe...",68430,53921300,https://www.imdb.com/title/tt2109184/
9997,9998,Dark Crimes,4.6,92 min,"Crime, Drama, Mystery",24.0,A murder investigation of a slain businessman ...,"['Alexandros Avranas', 'Jim Carrey', 'Marton C...","['Jim Carrey', 'Marton Csokas', 'Charlotte Gai...",19098,19098,https://www.imdb.com/title/tt1901024/
9998,9999,The Pyramid,4.6,89 min,"Horror, Mystery, Thriller",24.0,An archaeological team attempts to unlock the ...,"['Grégory Levasseur', 'Ashley Hinshaw', 'James...","['Ashley Hinshaw', 'James Buckley', ""Denis O'H...",25483,2756333,https://www.imdb.com/title/tt2799166/


In this `DataFrame` the index is an integer ranging from 0 to 9998 (since the data contains 9999) movies. We can change the index to be either the `ID` column of the dataset or the `Title` column

In [13]:
mv_idx = movies_df.set_index("ID")
mv_idx.head(3)

,title,Rating,Runtime,Genre,Metascore,Plot,Directors,Stars,Votes,Gross,Link
ID,,,,,,,,,,,
1,The Shawshank Redemption,9.3,142 min,Drama,82.0,"Over the course of several years, two convicts...","['Frank Darabont', 'Tim Robbins', 'Morgan Free...","['Tim Robbins', 'Morgan Freeman', 'Bob Gunton'...",2752419,28341469,https://www.imdb.com/title/tt0111161/
2,The Godfather,9.2,175 min,"Crime, Drama",100.0,"Don Vito Corleone, head of a mafia family, dec...","['Francis Ford Coppola', 'Marlon Brando', 'Al ...","['Marlon Brando', 'Al Pacino', 'James Caan', '...",1914751,134966411,https://www.imdb.com/title/tt0068646/
3,Ramayana: The Legend of Prince Rama,9.2,135 min,"Animation, Action, Adventure",NaN,An anime adaptation of the Hindu epic the Rama...,"['Ram Mohan', 'Yûgô Sakô', 'Koichi Saski', 'Ar...","['Yûgô Sakô', 'Koichi Saski', 'Arun Govil', 'N...",10108,10108,https://www.imdb.com/title/tt0259534/


In [14]:
mv_ttl = movies_df.set_index("title")
mv_ttl.head(3)

,ID,Rating,Runtime,Genre,Metascore,Plot,Directors,Stars,Votes,Gross,Link
title,,,,,,,,,,,
The Shawshank Redemption,1,9.3,142 min,Drama,82.0,"Over the course of several years, two convicts...","['Frank Darabont', 'Tim Robbins', 'Morgan Free...","['Tim Robbins', 'Morgan Freeman', 'Bob Gunton'...",2752419,28341469,https://www.imdb.com/title/tt0111161/
The Godfather,2,9.2,175 min,"Crime, Drama",100.0,"Don Vito Corleone, head of a mafia family, dec...","['Francis Ford Coppola', 'Marlon Brando', 'Al ...","['Marlon Brando', 'Al Pacino', 'James Caan', '...",1914751,134966411,https://www.imdb.com/title/tt0068646/
Ramayana: The Legend of Prince Rama,3,9.2,135 min,"Animation, Action, Adventure",NaN,An anime adaptation of the Hindu epic the Rama...,"['Ram Mohan', 'Yûgô Sakô', 'Koichi Saski', 'Ar...","['Yûgô Sakô', 'Koichi Saski', 'Arun Govil', 'N...",10108,10108,https://www.imdb.com/title/tt0259534/


## Handling missing values

First, we have check which columns have missing values. The `.isnull()` method returns a `DataFrame` containing the same number of rows and columns as the original `DataFrame` but each cell is either `True` or `False` indicating whether the data has missing value in the corresponding cell.

In [15]:
movies_df.isnull()

,ID,title,Rating,Runtime,Genre,Metascore,Plot,Directors,Stars,Votes,Gross,Link
0,False,False,False,False,False,False,False,False,False,False,False,False
1,False,False,False,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,True,False,False,False,False,False,False
3,False,False,False,False,False,True,False,False,False,False,False,False
4,False,False,False,False,False,True,False,False,False,False,False,False
...,...,...,...,...,...,...,...,...,...,...,...,...
9994,False,False,False,False,False,False,False,False,False,False,False,False
9995,False,False,False,False,False,False,False,False,False,False,False,False
9996,False,False,False,False,False,False,False,False,False,False,False,False
9997,False,False,False,False,False,False,False,False,False,False,False,False


In [16]:
# we can sum the output of .isnull() to get count of missing values
movies_df.isnull().sum()

,0
ID,0
title,0
Rating,0
Runtime,0
Genre,0
Metascore,2007
Plot,0
Directors,0
Stars,0
Votes,0


The `Metascore` column has some missing values.

### Complete case analysis

We can drop the missing rows containing values and perform further analysis with only the complete observations (this is known as complete case analysis).

In [17]:
movies_cc = movies_df.dropna()
movies_cc.isna().sum()    # the .isna() method can also be used

,0
ID,0
title,0
Rating,0
Runtime,0
Genre,0
Metascore,0
Plot,0
Directors,0
Stars,0
Votes,0


### Imputation
Imputation means filling out the missing values in a dataset with plausible values based on some criteria/logic/algorithm. There are many basic and advanced forms of imputation, however, they are out of the scope of this course. Therefore, we shall limit discussion on imputation at a minimum.

Missing values can be filled with the `.fillna()` method.

In [18]:
# replacing the missing values with the overall average value
movies_df['Metascore'] = movies_df['Metascore'].fillna(movies_df['Metascore'].mean())

In [19]:
movies_df['Metascore'].isnull().sum()    # missing count 0 now

np.int64(0)

## Indexing and Slicing a `DataFrame`

### Selecting a single column

In [20]:
genre_col = movies_df['Genre']

In [21]:
genre_col

,Genre
0,Drama
1,"Crime, Drama"
2,"Animation, Action, Adventure"
3,"Comedy, Drama"
4,"Adventure, Drama"
...,...
9994,"Animation, Adventure, Comedy"
9995,"Comedy, Fantasy, Horror"
9996,"Horror, Mystery"
9997,"Crime, Drama, Mystery"


In [22]:
print(type(genre_col))

<class 'pandas.core.series.Series'>


In [23]:
print(type(movies_df.Genre))    # same as above

<class 'pandas.core.series.Series'>


We can see above that selecting a single column with the `df['col_name']` syntax or the `df.col_name` syntax returns a `Series` object.

### Selecting multiple columns

We can also select one or more columns with `df[['col1', 'col2',...]` syntax. In this case the return value would of type `DataFrame` (even if we select only one column)

In [24]:
movies_df[['title', 'Runtime']].head(4)

,title,Runtime
0,The Shawshank Redemption,142 min
1,The Godfather,175 min
2,Ramayana: The Legend of Prince Rama,135 min
3,The Chaos Class,87 min


In [25]:
print(type(movies_df[['title', 'Runtime']]))

<class 'pandas.core.frame.DataFrame'>


In [26]:
print(type(movies_df[['title']]))

<class 'pandas.core.frame.DataFrame'>


### Selecting Columns Based on Data Types

Below we select the numeric columns:

In [27]:
movies_num = movies_df.select_dtypes("number")
movies_num.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9999 entries, 0 to 9998
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   ID         9999 non-null   int64  
 1   Rating     9999 non-null   float64
 2   Metascore  9999 non-null   float64
 3   Votes      9999 non-null   int64  
 4   Gross      9999 non-null   int64  
dtypes: float64(2), int64(3)
memory usage: 390.7 KB


### Selecting Rows

For slicing DataFrames by rows, we have two options
- `.loc` - locates by name or label
- `.iloc` - locates by numerical index

Earlier we defined `mv_ttl` which is the same as `movies_df` but with the `index` set to the `title` column.


#### Using `.iloc`

In [28]:
mv_ttl.iloc[0:4]

,ID,Rating,Runtime,Genre,Metascore,Plot,Directors,Stars,Votes,Gross,Link
title,,,,,,,,,,,
The Shawshank Redemption,1,9.3,142 min,Drama,82.0,"Over the course of several years, two convicts...","['Frank Darabont', 'Tim Robbins', 'Morgan Free...","['Tim Robbins', 'Morgan Freeman', 'Bob Gunton'...",2752419,28341469,https://www.imdb.com/title/tt0111161/
The Godfather,2,9.2,175 min,"Crime, Drama",100.0,"Don Vito Corleone, head of a mafia family, dec...","['Francis Ford Coppola', 'Marlon Brando', 'Al ...","['Marlon Brando', 'Al Pacino', 'James Caan', '...",1914751,134966411,https://www.imdb.com/title/tt0068646/
Ramayana: The Legend of Prince Rama,3,9.2,135 min,"Animation, Action, Adventure",NaN,An anime adaptation of the Hindu epic the Rama...,"['Ram Mohan', 'Yûgô Sakô', 'Koichi Saski', 'Ar...","['Yûgô Sakô', 'Koichi Saski', 'Arun Govil', 'N...",10108,10108,https://www.imdb.com/title/tt0259534/
The Chaos Class,4,9.2,87 min,"Comedy, Drama",NaN,"Lazy, uneducated students share a very close b...","['Ertem Egilmez', 'Kemal Sunal', 'Münir Özkul'...","['Kemal Sunal', 'Münir Özkul', 'Halit Akçatepe...",41785,41785,https://www.imdb.com/title/tt0252487/


In `.iloc` we can also mention specific columns to select. Inside the brackets, the indices before the comma specifies rows and indices after the comma specifies columns:

In [29]:
mv_ttl.iloc[0:4, 1:3]    # First four rows, 2nd and 3rd columns

,Rating,Runtime
title,,
The Shawshank Redemption,9.3,142 min
The Godfather,9.2,175 min
Ramayana: The Legend of Prince Rama,9.2,135 min
The Chaos Class,9.2,87 min


#### Using `.loc`

In [30]:
mv_ttl.loc["The Shawshank Redemption":"The Chaos Class", "Rating":"Runtime"]

,Rating,Runtime
title,,
The Shawshank Redemption,9.3,142 min
The Godfather,9.2,175 min
Ramayana: The Legend of Prince Rama,9.2,135 min
The Chaos Class,9.2,87 min


#### Inconsistency between `.loc` and `.iloc`

When slicing with `.iloc`, the last index in the slice is *NOT included* in the output. However, when indexing with `.loc`, the last index is *included* in the output. See the above outputs for an example.

## Filtering Values (selecting rows based on conditions)

Selecting movies with a high rating:

In [31]:
movies_df[movies_df['Rating'] > 8]

,ID,title,Rating,Runtime,Genre,Metascore,Plot,Directors,Stars,Votes,Gross,Link
0,1,The Shawshank Redemption,9.3,142 min,Drama,82.000000,"Over the course of several years, two convicts...","['Frank Darabont', 'Tim Robbins', 'Morgan Free...","['Tim Robbins', 'Morgan Freeman', 'Bob Gunton'...",2752419,28341469,https://www.imdb.com/title/tt0111161/
1,2,The Godfather,9.2,175 min,"Crime, Drama",100.000000,"Don Vito Corleone, head of a mafia family, dec...","['Francis Ford Coppola', 'Marlon Brando', 'Al ...","['Marlon Brando', 'Al Pacino', 'James Caan', '...",1914751,134966411,https://www.imdb.com/title/tt0068646/
2,3,Ramayana: The Legend of Prince Rama,9.2,135 min,"Animation, Action, Adventure",58.749625,An anime adaptation of the Hindu epic the Rama...,"['Ram Mohan', 'Yûgô Sakô', 'Koichi Saski', 'Ar...","['Yûgô Sakô', 'Koichi Saski', 'Arun Govil', 'N...",10108,10108,https://www.imdb.com/title/tt0259534/
3,4,The Chaos Class,9.2,87 min,"Comedy, Drama",58.749625,"Lazy, uneducated students share a very close b...","['Ertem Egilmez', 'Kemal Sunal', 'Münir Özkul'...","['Kemal Sunal', 'Münir Özkul', 'Halit Akçatepe...",41785,41785,https://www.imdb.com/title/tt0252487/
4,5,Daman,9.1,121 min,"Adventure, Drama",58.749625,"The film is set in 2015. Sid, is a young docto...","['Lenka Debiprasad', 'Vishal Mourya', 'Karan K...","['Vishal Mourya', 'Karan Kandhapan', 'Babushan...",13303,13303,https://www.imdb.com/title/tt17592606/
...,...,...,...,...,...,...,...,...,...,...,...,...
491,492,Anand,8.1,122 min,"Drama, Musical",58.749625,The story of a terminally ill man who wishes t...,"['Hrishikesh Mukherjee', 'Rajesh Khanna', 'Ami...","['Rajesh Khanna', 'Amitabh Bachchan', 'Sumita ...",34462,34462,https://www.imdb.com/title/tt0066763/
492,493,Iqbal,8.1,132 min,"Drama, Sport",58.749625,"Iqbal, the deaf and mute son of a farmer, has ...","['Nagesh Kukunoor', 'Shreyas Talpade', 'Naseer...","['Shreyas Talpade', 'Naseeruddin Shah', 'Shwet...",17329,17329,https://www.imdb.com/title/tt0453729/
493,494,The Legend of Bhagat Singh,8.1,155 min,"Action, Biography, Drama",58.749625,The story of a young revolutionary who raised ...,"['Rajkumar Santoshi', 'Ajay Devgn', 'Sushant S...","['Ajay Devgn', 'Sushant Singh', 'D. Santosh', ...",16881,16881,https://www.imdb.com/title/tt0319736/
494,495,Ghilli,8.1,160 min,"Action, Drama, Romance",58.749625,"Velu, an aspiring kabaddi player, goes to Madu...","['Dharani', 'Joseph Vijay', 'Trisha Krishnan',...","['Joseph Vijay', 'Trisha Krishnan', 'Prakash R...",15531,15531,https://www.imdb.com/title/tt0422320/


Sometimes a high rating alone may not be indicative of a good movie, so we can filter with a high number of votes too:

In [32]:
movies_df[(movies_df['Rating'] > 8) & (movies_df['Votes'] > 100000)]

,ID,title,Rating,Runtime,Genre,Metascore,Plot,Directors,Stars,Votes,Gross,Link
0,1,The Shawshank Redemption,9.3,142 min,Drama,82.000000,"Over the course of several years, two convicts...","['Frank Darabont', 'Tim Robbins', 'Morgan Free...","['Tim Robbins', 'Morgan Freeman', 'Bob Gunton'...",2752419,28341469,https://www.imdb.com/title/tt0111161/
1,2,The Godfather,9.2,175 min,"Crime, Drama",100.000000,"Don Vito Corleone, head of a mafia family, dec...","['Francis Ford Coppola', 'Marlon Brando', 'Al ...","['Marlon Brando', 'Al Pacino', 'James Caan', '...",1914751,134966411,https://www.imdb.com/title/tt0068646/
5,6,Spider-Man: Across the Spider-Verse,9.0,140 min,"Animation, Action, Adventure",86.000000,"Miles Morales catapults across the Multiverse,...","['Joaquim Dos Santos', 'Kemp Powers', 'Justin ...","['Kemp Powers', 'Justin K. Thompson', 'Shameik...",122030,122030,https://www.imdb.com/title/tt9362722/
6,7,The Dark Knight,9.0,152 min,"Action, Crime, Drama",84.000000,When the menace known as the Joker wreaks havo...,"['Christopher Nolan', 'Christian Bale', 'Heath...","['Christian Bale', 'Heath Ledger', 'Aaron Eckh...",2725188,534858444,https://www.imdb.com/title/tt0468569/
7,8,Schindler's List,9.0,195 min,"Biography, Drama, History",95.000000,"In German-occupied Poland during World War II,...","['Steven Spielberg', 'Liam Neeson', 'Ralph Fie...","['Liam Neeson', 'Ralph Fiennes', 'Ben Kingsley...",1387831,96898818,https://www.imdb.com/title/tt0108052/
...,...,...,...,...,...,...,...,...,...,...,...,...
408,409,Wild Strawberries,8.1,91 min,"Drama, Romance",88.000000,"After living a life marked by coldness, an agi...","['Ingmar Bergman', 'Victor Sjöström', 'Bibi An...","['Victor Sjöström', 'Bibi Andersson', 'Ingrid ...",110195,110195,https://www.imdb.com/title/tt0050986/
417,418,Lagaan: Once Upon a Time in India,8.1,224 min,"Drama, Musical, Sport",84.000000,The people of a small village in Victorian Ind...,"['Ashutosh Gowariker', 'Aamir Khan', 'Raghubir...","['Aamir Khan', 'Raghubir Yadav', 'Gracy Singh'...",117077,70147,https://www.imdb.com/title/tt0169102/
419,420,Mr. Smith Goes to Washington,8.1,129 min,"Comedy, Drama",73.000000,A naive youth leader is appointed to fill a va...,"['Frank Capra', 'James Stewart', 'Jean Arthur'...","['James Stewart', 'Jean Arthur', 'Claude Rains...",117774,9600000,https://www.imdb.com/title/tt0031679/
432,433,The Gold Rush,8.1,95 min,"Adventure, Comedy, Drama",58.749625,A prospector goes to the Klondike during the 1...,"['Charles Chaplin', 'Charles Chaplin', 'Mack S...","['Charles Chaplin', 'Mack Swain', 'Tom Murray'...",114081,5450000,https://www.imdb.com/title/tt0015864/


Filtering movies which have only "Action" in their genre:

In [33]:
movies_df[movies_df["Genre"] == "Action"]

,ID,title,Rating,Runtime,Genre,Metascore,Plot,Directors,Stars,Votes,Gross,Link
85,86,Banshee,8.5,50 min,Action,58.749625,A thief gets out of prison after many years an...,"['Gemma Mc Carthy', 'Antony Starr', ""Jonathan ...","['Antony Starr', ""Jonathan O' Dwyer"", 'Sean Fl...",11095,11095,https://www.imdb.com/title/tt3396910/
1684,1685,Ayan,7.6,162 min,Action,58.749625,A young man's smuggling lifestyle takes a seri...,"['K.V. Anand', 'Suriya', 'Tamannaah Bhatia', '...","['Suriya', 'Tamannaah Bhatia', 'Prabhu', 'Jagan']",11086,11086,https://www.imdb.com/title/tt1355546/
7284,7285,Ong Bak 2,6.2,98 min,Action,47.000000,A young Thai boxer learns the skills and inner...,"['Tony Jaa', 'Panna Rittikrai', 'Tony Jaa', 'S...","['Panna Rittikrai', 'Tony Jaa', 'Sarunyu Wongk...",31338,102055,https://www.imdb.com/title/tt0785035/
7707,7708,Velayudham,6.1,186 min,Action,58.749625,After a villager unwittingly prevents a number...,"['Mohan Raja', 'Joseph Vijay', ""Genelia D'Souz...","['Joseph Vijay', ""Genelia D'Souza"", 'Hansika M...",12375,12375,https://www.imdb.com/title/tt1965272/
9238,9239,The Man with the Iron Fists,5.4,95 min,Action,51.000000,"On the hunt for a fabled treasure of gold, a b...","['RZA', 'Russell Crowe', 'Cung Le', 'Lucy Liu'...","['Russell Crowe', 'Cung Le', 'Lucy Liu', 'RZA']",63551,15634090,https://www.imdb.com/title/tt1258972/
9825,9826,Ong Bak 3,4.9,99 min,Action,38.000000,"Continuing from Ong Bak 2's 1440s, Prince Tien...","['Tony Jaa', 'Panna Rittikrai', 'Tony Jaa', 'D...","['Panna Rittikrai', 'Tony Jaa', 'Dan Chupong',...",14331,14498,https://www.imdb.com/title/tt1653690/


Filtering movies which have "Action" or "Comedy" in their genre:

In [34]:
movies_df[(movies_df["Genre"] == "Action") | (movies_df["Genre"] == "Comedy")]

,ID,title,Rating,Runtime,Genre,Metascore,Plot,Directors,Stars,Votes,Gross,Link
14,15,Tosun Pasa,8.9,90 min,Comedy,58.749625,"Two traditionally rival Turkish families, ""Sef...","['Kartal Tibet', 'Kemal Sunal', 'Müjde Ar', 'A...","['Kemal Sunal', 'Müjde Ar', 'Adile Nasit', 'Se...",23933,23933,https://www.imdb.com/title/tt0253828/
15,16,The Chaos Class Failed the Class,8.9,91 min,Comedy,58.749625,A young and beautiful female teacher starts wo...,"['Ertem Egilmez', 'Kemal Sunal', 'Münir Özkul'...","['Kemal Sunal', 'Münir Özkul', 'Semra Özdamar'...",23976,23976,https://www.imdb.com/title/tt0252488/
85,86,Banshee,8.5,50 min,Action,58.749625,A thief gets out of prison after many years an...,"['Gemma Mc Carthy', 'Antony Starr', ""Jonathan ...","['Antony Starr', ""Jonathan O' Dwyer"", 'Sean Fl...",11095,11095,https://www.imdb.com/title/tt3396910/
140,141,"When I Grow Up, I'll Be a Kangaroo",8.4,92 min,Comedy,58.749625,"During one peculiar night, the lives of severa...","['Radivoje Andric', 'Sergej Trifunovic', 'Mari...","['Sergej Trifunovic', 'Marija Karan', 'Nebojsa...",11116,11116,https://www.imdb.com/title/tt0383846/
520,521,Young Frankenstein,8.0,106 min,Comedy,83.000000,An American grandson of the infamous scientist...,"['Mel Brooks', 'Gene Wilder', 'Madeline Kahn',...","['Gene Wilder', 'Madeline Kahn', 'Marty Feldma...",163092,86300000,https://www.imdb.com/title/tt0072431/
...,...,...,...,...,...,...,...,...,...,...,...,...
9916,9917,Recep Ivedik,4.8,90 min,Comedy,58.749625,"A man trying to impress his childhood lover, a...","['Togan Gökbakar', 'Sahan Gökbakar', 'Fatma To...","['Sahan Gökbakar', 'Fatma Toptas', 'Tulug Çizg...",27759,27759,https://www.imdb.com/title/tt1193516/
9921,9922,The Bubble,4.7,126 min,Comedy,34.000000,A group of actors and actresses stuck inside a...,"['Judd Apatow', 'Harry Trevaldwyn', 'Samson Ka...","['Harry Trevaldwyn', 'Samson Kayo', 'Peter Ser...",26922,26922,https://www.imdb.com/title/tt13610562/
9954,9955,Deuce Bigalow: European Gigolo,4.7,83 min,Comedy,23.000000,Deuce is tricked again into man-whoring by T.J...,"['Mike Bigelow', 'Rob Schneider', 'Eddie Griff...","['Rob Schneider', 'Eddie Griffin', 'Jeroen Kra...",52178,22264487,https://www.imdb.com/title/tt0367652/
9961,9962,Like a Boss,4.7,83 min,Comedy,33.000000,Two friends with very different ideals start a...,"['Miguel Arteta', 'Tiffany Haddish', 'Rose Byr...","['Tiffany Haddish', 'Rose Byrne', 'Salma Hayek...",14781,14781,https://www.imdb.com/title/tt7545266/


The above query can be concisely written using the `.isin()` method:

In [35]:
movies_df[movies_df["Genre"].isin(["Action", "Comedy"])]

,ID,title,Rating,Runtime,Genre,Metascore,Plot,Directors,Stars,Votes,Gross,Link
14,15,Tosun Pasa,8.9,90 min,Comedy,58.749625,"Two traditionally rival Turkish families, ""Sef...","['Kartal Tibet', 'Kemal Sunal', 'Müjde Ar', 'A...","['Kemal Sunal', 'Müjde Ar', 'Adile Nasit', 'Se...",23933,23933,https://www.imdb.com/title/tt0253828/
15,16,The Chaos Class Failed the Class,8.9,91 min,Comedy,58.749625,A young and beautiful female teacher starts wo...,"['Ertem Egilmez', 'Kemal Sunal', 'Münir Özkul'...","['Kemal Sunal', 'Münir Özkul', 'Semra Özdamar'...",23976,23976,https://www.imdb.com/title/tt0252488/
85,86,Banshee,8.5,50 min,Action,58.749625,A thief gets out of prison after many years an...,"['Gemma Mc Carthy', 'Antony Starr', ""Jonathan ...","['Antony Starr', ""Jonathan O' Dwyer"", 'Sean Fl...",11095,11095,https://www.imdb.com/title/tt3396910/
140,141,"When I Grow Up, I'll Be a Kangaroo",8.4,92 min,Comedy,58.749625,"During one peculiar night, the lives of severa...","['Radivoje Andric', 'Sergej Trifunovic', 'Mari...","['Sergej Trifunovic', 'Marija Karan', 'Nebojsa...",11116,11116,https://www.imdb.com/title/tt0383846/
520,521,Young Frankenstein,8.0,106 min,Comedy,83.000000,An American grandson of the infamous scientist...,"['Mel Brooks', 'Gene Wilder', 'Madeline Kahn',...","['Gene Wilder', 'Madeline Kahn', 'Marty Feldma...",163092,86300000,https://www.imdb.com/title/tt0072431/
...,...,...,...,...,...,...,...,...,...,...,...,...
9916,9917,Recep Ivedik,4.8,90 min,Comedy,58.749625,"A man trying to impress his childhood lover, a...","['Togan Gökbakar', 'Sahan Gökbakar', 'Fatma To...","['Sahan Gökbakar', 'Fatma Toptas', 'Tulug Çizg...",27759,27759,https://www.imdb.com/title/tt1193516/
9921,9922,The Bubble,4.7,126 min,Comedy,34.000000,A group of actors and actresses stuck inside a...,"['Judd Apatow', 'Harry Trevaldwyn', 'Samson Ka...","['Harry Trevaldwyn', 'Samson Kayo', 'Peter Ser...",26922,26922,https://www.imdb.com/title/tt13610562/
9954,9955,Deuce Bigalow: European Gigolo,4.7,83 min,Comedy,23.000000,Deuce is tricked again into man-whoring by T.J...,"['Mike Bigelow', 'Rob Schneider', 'Eddie Griff...","['Rob Schneider', 'Eddie Griffin', 'Jeroen Kra...",52178,22264487,https://www.imdb.com/title/tt0367652/
9961,9962,Like a Boss,4.7,83 min,Comedy,33.000000,Two friends with very different ideals start a...,"['Miguel Arteta', 'Tiffany Haddish', 'Rose Byr...","['Tiffany Haddish', 'Rose Byrne', 'Salma Hayek...",14781,14781,https://www.imdb.com/title/tt7545266/


Selecting movies with gross revenue above the overall 75% percentile of all movies in the datset:

In [36]:
movies_df[movies_df['Gross'] > movies_df['Gross'].quantile(0.75)]

,ID,title,Rating,Runtime,Genre,Metascore,Plot,Directors,Stars,Votes,Gross,Link
1,2,The Godfather,9.2,175 min,"Crime, Drama",100.0,"Don Vito Corleone, head of a mafia family, dec...","['Francis Ford Coppola', 'Marlon Brando', 'Al ...","['Marlon Brando', 'Al Pacino', 'James Caan', '...",1914751,134966411,https://www.imdb.com/title/tt0068646/
6,7,The Dark Knight,9.0,152 min,"Action, Crime, Drama",84.0,When the menace known as the Joker wreaks havo...,"['Christopher Nolan', 'Christian Bale', 'Heath...","['Christian Bale', 'Heath Ledger', 'Aaron Eckh...",2725188,534858444,https://www.imdb.com/title/tt0468569/
7,8,Schindler's List,9.0,195 min,"Biography, Drama, History",95.0,"In German-occupied Poland during World War II,...","['Steven Spielberg', 'Liam Neeson', 'Ralph Fie...","['Liam Neeson', 'Ralph Fiennes', 'Ben Kingsley...",1387831,96898818,https://www.imdb.com/title/tt0108052/
9,10,The Godfather Part II,9.0,202 min,"Crime, Drama",90.0,The early life and career of Vito Corleone in ...,"['Francis Ford Coppola', 'Al Pacino', 'Robert ...","['Al Pacino', 'Robert De Niro', 'Robert Duvall...",1303858,57300000,https://www.imdb.com/title/tt0071562/
10,11,The Lord of the Rings: The Return of the King,9.0,201 min,"Action, Adventure, Drama",94.0,Gandalf and Aragorn lead the World of Men agai...,"['Peter Jackson', 'Elijah Wood', 'Viggo Morten...","['Elijah Wood', 'Viggo Mortensen', 'Ian McKell...",1890885,377845905,https://www.imdb.com/title/tt0167260/
...,...,...,...,...,...,...,...,...,...,...,...,...
9988,9989,Aliens vs. Predator: Requiem,4.6,94 min,"Action, Adventure, Horror",29.0,Warring Alien and Predator races descend on a ...,"['Colin Strause', 'Greg Strause', 'Reiko Ayles...","['Greg Strause', 'Reiko Aylesworth', 'Steven P...",129456,41797066,https://www.imdb.com/title/tt0758730/
9990,9991,Hannah Montana: The Movie,4.6,102 min,"Comedy, Drama, Family",47.0,As Hannah Montana's popularity begins to take ...,"['Peter Chelsom', 'Miley Cyrus', 'Emily Osment...","['Miley Cyrus', 'Emily Osment', 'Billy Ray Cyr...",44174,79566871,https://www.imdb.com/title/tt1114677/
9991,9992,God's Not Dead,4.6,113 min,"Adventure, Comedy, Drama",16.0,College philosophy professor Mr. Radisson's cu...,"['Harold Cronk', 'Shane Harper', 'Kevin Sorbo'...","['Shane Harper', 'Kevin Sorbo', 'David A.R. Wh...",43934,60755732,https://www.imdb.com/title/tt2528814/
9994,9995,Yogi Bear,4.6,80 min,"Animation, Adventure, Comedy",35.0,A documentary filmmaker travels to Jellystone ...,"['Eric Brevig', 'Dan Aykroyd', 'Justin Timberl...","['Dan Aykroyd', 'Justin Timberlake', 'Anna Far...",22762,100246011,https://www.imdb.com/title/tt1302067/


In [37]:
# filtering movies with rating between 8 and 9:
movies_df[(movies_df["Rating"] >= 8) & (movies_df["Rating"] <= 9)]

,ID,title,Rating,Runtime,Genre,Metascore,Plot,Directors,Stars,Votes,Gross,Link
5,6,Spider-Man: Across the Spider-Verse,9.0,140 min,"Animation, Action, Adventure",86.000000,"Miles Morales catapults across the Multiverse,...","['Joaquim Dos Santos', 'Kemp Powers', 'Justin ...","['Kemp Powers', 'Justin K. Thompson', 'Shameik...",122030,122030,https://www.imdb.com/title/tt9362722/
6,7,The Dark Knight,9.0,152 min,"Action, Crime, Drama",84.000000,When the menace known as the Joker wreaks havo...,"['Christopher Nolan', 'Christian Bale', 'Heath...","['Christian Bale', 'Heath Ledger', 'Aaron Eckh...",2725188,534858444,https://www.imdb.com/title/tt0468569/
7,8,Schindler's List,9.0,195 min,"Biography, Drama, History",95.000000,"In German-occupied Poland during World War II,...","['Steven Spielberg', 'Liam Neeson', 'Ralph Fie...","['Liam Neeson', 'Ralph Fiennes', 'Ben Kingsley...",1387831,96898818,https://www.imdb.com/title/tt0108052/
8,9,12 Angry Men,9.0,96 min,"Crime, Drama",97.000000,The jury in a New York City murder trial is fr...,"['Sidney Lumet', 'Henry Fonda', 'Lee J. Cobb',...","['Henry Fonda', 'Lee J. Cobb', 'Martin Balsam'...",815205,4360000,https://www.imdb.com/title/tt0050083/
9,10,The Godfather Part II,9.0,202 min,"Crime, Drama",90.000000,The early life and career of Vito Corleone in ...,"['Francis Ford Coppola', 'Al Pacino', 'Robert ...","['Al Pacino', 'Robert De Niro', 'Robert Duvall...",1303858,57300000,https://www.imdb.com/title/tt0071562/
...,...,...,...,...,...,...,...,...,...,...,...,...
670,671,My Grandfather's People,8.0,126 min,"Comedy, Drama, Family",58.749625,Grandpapa and his family were torn from their ...,"['Çagan Irmak', 'Çetin Tekindor', 'Yigit Özsen...","['Çetin Tekindor', 'Yigit Özsener', 'Gökçe Bah...",11023,11023,https://www.imdb.com/title/tt2150209/
671,672,Manjhi: The Mountain Man,8.0,120 min,"Biography, Drama",58.749625,"Using only a hammer and a chisel, a man spends...","['Ketan Mehta', 'Nawazuddin Siddiqui', 'Abhina...","['Nawazuddin Siddiqui', 'Abhinay Raj Singh', '...",19577,19577,https://www.imdb.com/title/tt3449292/
672,673,Pokiri,8.0,164 min,"Action, Crime, Thriller",58.749625,A money-minded criminal has to deal with two r...,"['Puri Jagannadh', 'Mahesh Babu', ""Ileana D'Cr...","['Mahesh Babu', ""Ileana D'Cruz"", 'Prakash Raj'...",15030,15030,https://www.imdb.com/title/tt0816258/
673,674,Cars of the Revolution,8.0,115 min,Drama,58.749625,"In the years of military coup, Turkey was unab...","['Tolga Örnek', 'Haluk Bilginer', 'Taner Birse...","['Haluk Bilginer', 'Taner Birsel', 'Charles Ca...",12416,12416,https://www.imdb.com/title/tt1282139/


The above can be rewritten more concisely using the `.between()` method:

In [38]:
movies_df[movies_df["Rating"].between(8, 9)]

,ID,title,Rating,Runtime,Genre,Metascore,Plot,Directors,Stars,Votes,Gross,Link
5,6,Spider-Man: Across the Spider-Verse,9.0,140 min,"Animation, Action, Adventure",86.000000,"Miles Morales catapults across the Multiverse,...","['Joaquim Dos Santos', 'Kemp Powers', 'Justin ...","['Kemp Powers', 'Justin K. Thompson', 'Shameik...",122030,122030,https://www.imdb.com/title/tt9362722/
6,7,The Dark Knight,9.0,152 min,"Action, Crime, Drama",84.000000,When the menace known as the Joker wreaks havo...,"['Christopher Nolan', 'Christian Bale', 'Heath...","['Christian Bale', 'Heath Ledger', 'Aaron Eckh...",2725188,534858444,https://www.imdb.com/title/tt0468569/
7,8,Schindler's List,9.0,195 min,"Biography, Drama, History",95.000000,"In German-occupied Poland during World War II,...","['Steven Spielberg', 'Liam Neeson', 'Ralph Fie...","['Liam Neeson', 'Ralph Fiennes', 'Ben Kingsley...",1387831,96898818,https://www.imdb.com/title/tt0108052/
8,9,12 Angry Men,9.0,96 min,"Crime, Drama",97.000000,The jury in a New York City murder trial is fr...,"['Sidney Lumet', 'Henry Fonda', 'Lee J. Cobb',...","['Henry Fonda', 'Lee J. Cobb', 'Martin Balsam'...",815205,4360000,https://www.imdb.com/title/tt0050083/
9,10,The Godfather Part II,9.0,202 min,"Crime, Drama",90.000000,The early life and career of Vito Corleone in ...,"['Francis Ford Coppola', 'Al Pacino', 'Robert ...","['Al Pacino', 'Robert De Niro', 'Robert Duvall...",1303858,57300000,https://www.imdb.com/title/tt0071562/
...,...,...,...,...,...,...,...,...,...,...,...,...
670,671,My Grandfather's People,8.0,126 min,"Comedy, Drama, Family",58.749625,Grandpapa and his family were torn from their ...,"['Çagan Irmak', 'Çetin Tekindor', 'Yigit Özsen...","['Çetin Tekindor', 'Yigit Özsener', 'Gökçe Bah...",11023,11023,https://www.imdb.com/title/tt2150209/
671,672,Manjhi: The Mountain Man,8.0,120 min,"Biography, Drama",58.749625,"Using only a hammer and a chisel, a man spends...","['Ketan Mehta', 'Nawazuddin Siddiqui', 'Abhina...","['Nawazuddin Siddiqui', 'Abhinay Raj Singh', '...",19577,19577,https://www.imdb.com/title/tt3449292/
672,673,Pokiri,8.0,164 min,"Action, Crime, Thriller",58.749625,A money-minded criminal has to deal with two r...,"['Puri Jagannadh', 'Mahesh Babu', ""Ileana D'Cr...","['Mahesh Babu', ""Ileana D'Cruz"", 'Prakash Raj'...",15030,15030,https://www.imdb.com/title/tt0816258/
673,674,Cars of the Revolution,8.0,115 min,Drama,58.749625,"In the years of military coup, Turkey was unab...","['Tolga Örnek', 'Haluk Bilginer', 'Taner Birse...","['Haluk Bilginer', 'Taner Birsel', 'Charles Ca...",12416,12416,https://www.imdb.com/title/tt1282139/


## Sorting Values

In [39]:
movies_df.sort_values(by = ["Rating"], ascending=True)

,ID,title,Rating,Runtime,Genre,Metascore,Plot,Directors,Stars,Votes,Gross,Link
9998,9999,The Pyramid,4.6,89 min,"Horror, Mystery, Thriller",24.000000,An archaeological team attempts to unlock the ...,"['Grégory Levasseur', 'Ashley Hinshaw', 'James...","['Ashley Hinshaw', 'James Buckley', ""Denis O'H...",25483,2756333,https://www.imdb.com/title/tt2799166/
9993,9994,Driven,4.6,116 min,"Action, Drama, Sport",29.000000,A young hot shot driver is in the middle of a ...,"['Renny Harlin', 'Sylvester Stallone', 'Kip Pa...","['Sylvester Stallone', 'Kip Pardue', 'Til Schw...",42226,32720065,https://www.imdb.com/title/tt0132245/
9992,9993,Psycho,4.6,105 min,"Horror, Mystery, Thriller",47.000000,A young female embezzler arrives at the Bates ...,"['Gus Van Sant', 'Vince Vaughn', 'Anne Heche',...","['Vince Vaughn', 'Anne Heche', 'Julianne Moore...",49548,21456130,https://www.imdb.com/title/tt0155975/
9994,9995,Yogi Bear,4.6,80 min,"Animation, Adventure, Comedy",35.000000,A documentary filmmaker travels to Jellystone ...,"['Eric Brevig', 'Dan Aykroyd', 'Justin Timberl...","['Dan Aykroyd', 'Justin Timberlake', 'Anna Far...",22762,100246011,https://www.imdb.com/title/tt1302067/
9997,9998,Dark Crimes,4.6,92 min,"Crime, Drama, Mystery",24.000000,A murder investigation of a slain businessman ...,"['Alexandros Avranas', 'Jim Carrey', 'Marton C...","['Jim Carrey', 'Marton Csokas', 'Charlotte Gai...",19098,19098,https://www.imdb.com/title/tt1901024/
...,...,...,...,...,...,...,...,...,...,...,...,...
4,5,Daman,9.1,121 min,"Adventure, Drama",58.749625,"The film is set in 2015. Sid, is a young docto...","['Lenka Debiprasad', 'Vishal Mourya', 'Karan K...","['Vishal Mourya', 'Karan Kandhapan', 'Babushan...",13303,13303,https://www.imdb.com/title/tt17592606/
2,3,Ramayana: The Legend of Prince Rama,9.2,135 min,"Animation, Action, Adventure",58.749625,An anime adaptation of the Hindu epic the Rama...,"['Ram Mohan', 'Yûgô Sakô', 'Koichi Saski', 'Ar...","['Yûgô Sakô', 'Koichi Saski', 'Arun Govil', 'N...",10108,10108,https://www.imdb.com/title/tt0259534/
3,4,The Chaos Class,9.2,87 min,"Comedy, Drama",58.749625,"Lazy, uneducated students share a very close b...","['Ertem Egilmez', 'Kemal Sunal', 'Münir Özkul'...","['Kemal Sunal', 'Münir Özkul', 'Halit Akçatepe...",41785,41785,https://www.imdb.com/title/tt0252487/
1,2,The Godfather,9.2,175 min,"Crime, Drama",100.000000,"Don Vito Corleone, head of a mafia family, dec...","['Francis Ford Coppola', 'Marlon Brando', 'Al ...","['Marlon Brando', 'Al Pacino', 'James Caan', '...",1914751,134966411,https://www.imdb.com/title/tt0068646/


In [40]:
movies_df.sort_values(by = ["Rating", "Votes"], ascending=[True, False])

,ID,title,Rating,Runtime,Genre,Metascore,Plot,Directors,Stars,Votes,Gross,Link
9988,9989,Aliens vs. Predator: Requiem,4.6,94 min,"Action, Adventure, Horror",29.000000,Warring Alien and Predator races descend on a ...,"['Colin Strause', 'Greg Strause', 'Reiko Ayles...","['Greg Strause', 'Reiko Aylesworth', 'Steven P...",129456,41797066,https://www.imdb.com/title/tt0758730/
9983,9984,Fifty Shades Darker,4.6,118 min,"Drama, Romance, Thriller",33.000000,While Christian wrestles with his inner demons...,"['James Foley', 'Dakota Johnson', 'Jamie Dorna...","['Dakota Johnson', 'Jamie Dornan', 'Eric Johns...",106578,114376885,https://www.imdb.com/title/tt4465564/
9985,9986,Death Note,4.6,101 min,"Crime, Drama, Fantasy",43.000000,A high school student named Light Turner disco...,"['Adam Wingard', 'Nat Wolff', 'LaKeith Stanfie...","['Nat Wolff', 'LaKeith Stanfield', 'Margaret Q...",87627,87627,https://www.imdb.com/title/tt1241317/
9984,9985,Superhero Movie,4.6,75 min,"Action, Comedy, Sci-Fi",33.000000,Orphaned high school student Rick Riker is bit...,"['Craig Mazin', 'Drake Bell', 'Leslie Nielsen'...","['Drake Bell', 'Leslie Nielsen', 'Sara Paxton'...",72271,26638520,https://www.imdb.com/title/tt0426592/
9996,9997,Paranormal Activity 4,4.6,88 min,"Horror, Mystery",40.000000,It has been five years since the disappearance...,"['Henry Joost', 'Ariel Schulman', 'Stephen Dun...","['Ariel Schulman', 'Stephen Dunham', 'Katie Fe...",68430,53921300,https://www.imdb.com/title/tt2109184/
...,...,...,...,...,...,...,...,...,...,...,...,...
4,5,Daman,9.1,121 min,"Adventure, Drama",58.749625,"The film is set in 2015. Sid, is a young docto...","['Lenka Debiprasad', 'Vishal Mourya', 'Karan K...","['Vishal Mourya', 'Karan Kandhapan', 'Babushan...",13303,13303,https://www.imdb.com/title/tt17592606/
1,2,The Godfather,9.2,175 min,"Crime, Drama",100.000000,"Don Vito Corleone, head of a mafia family, dec...","['Francis Ford Coppola', 'Marlon Brando', 'Al ...","['Marlon Brando', 'Al Pacino', 'James Caan', '...",1914751,134966411,https://www.imdb.com/title/tt0068646/
3,4,The Chaos Class,9.2,87 min,"Comedy, Drama",58.749625,"Lazy, uneducated students share a very close b...","['Ertem Egilmez', 'Kemal Sunal', 'Münir Özkul'...","['Kemal Sunal', 'Münir Özkul', 'Halit Akçatepe...",41785,41785,https://www.imdb.com/title/tt0252487/
2,3,Ramayana: The Legend of Prince Rama,9.2,135 min,"Animation, Action, Adventure",58.749625,An anime adaptation of the Hindu epic the Rama...,"['Ram Mohan', 'Yûgô Sakô', 'Koichi Saski', 'Ar...","['Yûgô Sakô', 'Koichi Saski', 'Arun Govil', 'N...",10108,10108,https://www.imdb.com/title/tt0259534/


## Data Exploration

### The `.describe()` Method

Using `describe()` on an entire `DataFrame` we can get a summary of the distribution of *continuous* variables

In [41]:
movies_df.describe()

,ID,Rating,Metascore,Votes,Gross
count,9999.000000,9999.000000,9999.000000,9.999000e+03,9.999000e+03
mean,5000.000000,6.698620,58.749625,9.127103e+04,2.874067e+07
std,2886.607005,0.852755,15.620576,1.684508e+05,5.971933e+07
min,1.000000,4.600000,7.000000,1.000100e+04,7.000000e+00
25%,2500.500000,6.100000,49.000000,1.678500e+04,3.654100e+04
50%,5000.000000,6.700000,58.749625,3.391200e+04,4.244155e+06
75%,7499.500000,7.300000,68.000000,8.965600e+04,3.306295e+07
max,9999.000000,9.300000,100.000000,2.752419e+06,9.366622e+08


It shows the non-null count, mean, standard deviation, min and max values, and the values of the quartiles of each continuous variable.

In the above, the summary of the `ID` column is irrelevant since it does not represent actual data.

The `Runtime` column contains the length of each movie in minutes, this should logically make the column `integer` type, however, each value is suffixed with the word `"min"`, therefore, the column is `string` type. We need to split the values of this column, take the first item after splitting and convert that to integer type.

In [42]:
movies_df['Runtime'] = movies_df['Runtime'].str.split().str[0].astype(int)

In [43]:
movies_df.drop('ID', axis = 1).describe()    # summary without the ID column

,Rating,Runtime,Metascore,Votes,Gross
count,9999.000000,9999.000000,9999.000000,9.999000e+03,9.999000e+03
mean,6.698620,110.479148,58.749625,9.127103e+04,2.874067e+07
std,0.852755,21.995584,15.620576,1.684508e+05,5.971933e+07
min,4.600000,45.000000,7.000000,1.000100e+04,7.000000e+00
25%,6.100000,96.000000,49.000000,1.678500e+04,3.654100e+04
50%,6.700000,106.000000,58.749625,3.391200e+04,4.244155e+06
75%,7.300000,120.000000,68.000000,8.965600e+04,3.306295e+07
max,9.300000,439.000000,100.000000,2.752419e+06,9.366622e+08


The `.describe()` method can also be run on individual non-numeric columns:

In [44]:
movies_df['Genre'].describe()

,Genre
count,9999
unique,428
top,"Comedy, Drama, Romance"
freq,480


It shows the count of the non-null values, the number of unique values, the value that appears the most and how many times it appears (`freq`).

### The `.value_counts()` method
For a categorical column, we can also run `.value_counts()`:

In [45]:
movies_df.Genre.value_counts()
# here, movies_df.Genre is the same as movies_df.['Genre']

,count
Genre,
"Comedy, Drama, Romance",480
Drama,433
"Comedy, Drama",408
"Drama, Romance",336
"Action, Crime, Drama",325
...,...
"Action, Thriller, Western",1
"Action, Drama, Music",1
"Comedy, Fantasy, Thriller",1


### The `.corr()` Method
Returns a correlation matrix

In [46]:
movies_df.drop('ID', axis = 1).corr(numeric_only=True)

,Rating,Runtime,Metascore,Votes,Gross
Rating,1.000000,0.345162,0.623556,0.307727,0.064945
Runtime,0.345162,1.000000,0.181856,0.197374,0.124496
Metascore,0.623556,0.181856,1.000000,0.202101,0.029370
Votes,0.307727,0.197374,0.202101,1.000000,0.632156
Gross,0.064945,0.124496,0.029370,0.632156,1.000000


## Computing New Columns

In [47]:
# creating a column depticcting gross revenue per minute of runtime
mv_idx['revenue_per_minute'] = movies_df['Gross'] / movies_df['Runtime']
mv_idx['revenue_per_minute'].head()

,revenue_per_minute
ID,
1,771236.634286
2,74.874074
3,480.287356
4,109.942149
5,871.642857


### The `.transform()` Method

This method applies a function to an axis (either column or row) and returns a `DataFrame` of the same shape.

We can mention `axis=1` for applying the function across the columns (row-wise), meaning that with `axis=1`, the function will get individual rows as input. Otherwise, it is column-wise.

For example, if we wanted to standardize the rating column:

In [48]:
mv_idx['rating_standardized'] = mv_idx['Rating'].transform(lambda x: (x-x.mean())/x.std())

In [49]:
mv_idx['rating_standardized'].head()

,rating_standardized
ID,
1,3.050559
2,2.933292
3,2.933292
4,2.933292
5,2.816025


In [50]:
mv_idx['rating_standardized'].mean()

np.float64(-1.3643784905084433e-16)

In [51]:
mv_idx['rating_standardized'].std()

1.0000000000000062

### The `.assign()` method

Assigns new columns to a `DataFrame`. Returns a new object with all original columns in addition to new ones. Existing columns that are re-assigned will be overwritten.

In [52]:
mv_idx = mv_idx.assign(
    rating_out_of_1 = mv_idx['Rating'] / 10,
    revenue_std = lambda df: (df['Gross'] - df['Gross'].mean())/df['Gross'].std()
)

mv_idx[['rating_out_of_1', 'revenue_std']].head()

,rating_out_of_1,revenue_std
ID,,
1,0.93,-0.006685
2,0.92,1.778750
3,0.92,-0.481093
4,0.92,-0.480563
5,0.91,-0.481040


### Discretization and Binning

Sometimes we need to numeric/continuous columns to categorical/discrete columns.

#### Using `pd.cut()`

This functions "cuts" continuous variable into several groups making it discrete or categorical.

```
pandas.cut(x, bins, right=True, labels=None, retbins=False, precision=3, include_lowest=False, duplicates='raise', ordered=True)
```

Parameters:
- x: array-like (series/column). The input array to be binned. Must be 1-dimensional.

- bins: int, sequence of scalars, or IntervalIndex. The criteria to bin by.
    - int : Defines the number of equal-width bins in the range of x. The range of x is extended by .1% on each side to include the minimum and maximum values of x.
    - sequence of scalars : Defines the bin edges allowing for non-uniform width. No extension of the range of x is done.
    - IntervalIndex : Defines the exact bins to be used. Note that IntervalIndex for bins must be non-overlapping.
- right: bool, default=True. Indicates whether bins includes the rightmost edge or not. If right == True (the default), then the bins [1, 2, 3, 4] indicate (1,2], (2,3], (3,4]. This argument is ignored when bins is an IntervalIndex.
- labels: array or False, default=None. Specifies the labels for the returned bins. Must be the same length as the resulting bins.
- include_lowest: bool, default=False. Whether the first interval should be left-inclusive or not.
- ordered: bool, default=True. Whether the labels are ordered or not. Applies to returned types Categorical and Series (with Categorical dtype). If True, the resulting categorical will be ordered. If False, the resulting categorical will be unordered (labels must be provided).



In [53]:
rating_cat = pd.cut(movies_df['Rating'], 3)
rating_cat.value_counts()

,count
Rating,
"(6.167, 7.733]",6255
"(4.595, 6.167]",2650
"(7.733, 9.3]",1094


Can mention `right = False` to include lower limits and not upper limits:

In [54]:
rating_cat2 = pd.cut(movies_df['Rating'], 3, right = False)
rating_cat2.value_counts()

,count
Rating,
"[6.167, 7.733)",6255
"[4.6, 6.167)",2650
"[7.733, 9.305)",1094


In [55]:
rating_cat3 = pd.cut(movies_df['Rating'], 3, include_lowest=True)
rating_cat3.value_counts()

,count
Rating,
"(6.167, 7.733]",6255
"(4.593999999999999, 6.167]",2650
"(7.733, 9.3]",1094


In [56]:
rating_cat4 = pd.cut(movies_df.Rating, [0, 5, 8, 10], right = True)
rating_cat4.value_counts()

,count
Rating,
"(5, 8]",9204
"(8, 10]",496
"(0, 5]",299


Can also provide a `list` of labels for each category using the `labels` argument:

In [57]:
rating_cat5 = pd.cut(movies_df.Rating, [0, 5, 8, 10], labels = ['<=5', '5-8', '>8'])
rating_cat5.value_counts()

,count
Rating,
5-8,9204
>8,496
<=5,299


#### `pd.qcut()`

Quantile-based discretization function. Uses sample quantiles to bin.

```
Parameters
----------
x : 1d ndarray or Series
q : int or list-like of float
    int: Number of quantiles. 10 for deciles, 4 for quartiles, etc.
    list: Alternately array of quantiles, e.g. [0, .25, .5, .75, 1.] for quartiles.
labels : array or False, default None
    Used as labels for the resulting bins. Must be of the same length as
    the resulting bins. If False, return only integer indicators of the
    bins. If True, raises an error.
retbins : bool, optional
    Whether to return the (bins, labels) or not. Can be useful if bins
    is given as a scalar.
```

In [58]:
pd.qcut(x = movies_df.Rating, q = 3).value_counts()

,count
Rating,
"(4.598999999999999, 6.3]",3441
"(6.3, 7.1]",3309
"(7.1, 9.3]",3249


In [59]:
pd.qcut(x = movies_df.Rating, q = [0, 0.25, 0.5, 0.75, 1]).value_counts()

,count
Rating,
"(4.598999999999999, 6.1]",2650
"(6.1, 6.7]",2476
"(6.7, 7.3]",2468
"(7.3, 9.3]",2405


### Aggregation

Aggregation functions are used to obtain summary information on columns of a dataframe using the `.agg()` method

#### Method 1: providing a callable function function to `.agg()`

In [60]:
import numpy as np

In [61]:
movies_df[["Gross", "Rating"]].agg(np.mean)

/tmp/ipython-input-2274451635.py:1: FutureWarning: The provided callable <function mean at 0x79daba578220> is currently using DataFrame.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  movies_df[["Gross", "Rating"]].agg(np.mean)


,0
Gross,2.874067e+07
Rating,6.698620e+00


Can also pass a lambda function:

In [62]:
movies_df[["Rating", "Gross"]].agg(lambda x: np.mean(x))

,0
Rating,6.698620e+00
Gross,2.874067e+07


#### Method 2: Selecting the columns, then providing the aggregation function name as strings to `.agg()` as a list of string:

In [63]:
movies_df[["Gross", "Rating"]].agg(["max", "min", "mean"])

,Gross,Rating
max,9.366622e+08,9.30000
min,7.000000e+00,4.60000
mean,2.874067e+07,6.69862


#### Method 3: Passing tuples of column name and aggregate function name as string

In [64]:
movies_df.agg(max_revenue = ("Gross", "max"), min_revenue = ("Gross", "min"), max_rating = ("Rating", "max"))

,Gross,Rating
max_revenue,936662225.0,NaN
min_revenue,7.0,NaN
max_rating,NaN,9.3


#### Method 4: Passing a dictionary with keys as column names on which to perform the aggregation, and lists of aggregating function names as strings as the values to the corresponding keys of the dictionary

In [65]:
movies_df.agg({
    "Rating" : ["mean", "min", "max"],
    "Gross" : ["min", "max", "std"]
})

,Rating,Gross
mean,6.69862,NaN
min,4.60000,7.000000e+00
max,9.30000,9.366622e+08
std,NaN,5.971933e+07


Below is a non-comprehensive list of `pandas` aggregation functions:

| Function      | Description |
|---------------|-------------|
| sum           | Adds up the values. |
| mean          | Computes the average of the values. |
| median        | Returns the median value. |
| min           | Gives the smallest value. |
| max           | Gives the largest value. |
| count         | Counts non-missing values. |
| size          | Counts all rows, including missing values. |
| nunique       | Counts distinct values. |
| std           | Computes standard deviation. |
| var           | Computes variance. |
| prod          | Multiplies all values. |
| first         | Returns the first non-missing value. |
| last          | Returns the last non-missing value. |
| idxmin        | Returns the index of the minimum value. |
| idxmax        | Returns the index of the maximum value. |
| quantile      | Returns a specified quantile. |
| mad           | Computes mean absolute deviation. |
| sem           | Computes standard error of the mean. |
| skew          | Measures distribution skewness. |
| kurt          | Measures distribution kurtosis. |
| corr          | Computes correlation with another series. |
| cov           | Computes covariance with another series. |
| mode          | Returns the mode(s) of the values. |
| any           | Tests if any element is True. |
| all           | Tests if all elements are True. |

Apart from the above, some `numpy` universal functions (`unfuncs`) can also be used.

### Applying Functions and Mapping


A frequent operation while analyzing a data is to apply a function on one-dimensional arrays to each column or each row. It can be done with the `.apply()` method of a `DataFrame`.

The `.apply()` method takes a function as its argument as well as the axis *across* which the function is applied. The function can be built-in, from a package or a lambda.

For example, we can define a function that returns the range and the 90th percentile of a variable:

In [66]:
def mysummary(x):
    return [x.max() - x.min(), x.quantile(0.90)]

If we want to find the range and the median of say, rating and gross revenue:

In [67]:

movies_df[['Rating', 'Gross']].apply(mysummary, axis=0)

,Rating,Gross
0,4.7,936662218.0
1,7.8,80572126.0


We can see that the two summary statistics returned from our defined function, namely, range and 90th percentile are indexed as 0 and 1, respectively in the output.

While outputting multiple summary statistics, it is desirable to index those statistics with appropriate labels.

We can provide appropriate labels to our outputs by returning a Series instead of a list as follows:

In [68]:
def mysummary(x):
    return pd.Series([x.max() - x.min(), x.quantile(0.90)],
                    index = ['range', '90th Percentile'])
movies_df[['Runtime', 'Gross']].apply(mysummary, axis=0)

,Runtime,Gross
range,394.0,936662218.0
90th Percentile,138.0,80572126.0


We can also invoke lambda functions inside the `.apply()` method instead of defining a function using the `def` keyword:

In [69]:
movies_df[['Runtime', 'Gross']].apply(
    lambda x: pd.Series(
        [x.max() - x.min(), x.quantile(0.90)],
        index = ['range', '90th Percentile']
    )
)

,Runtime,Gross
range,394.0,936662218.0
90th Percentile,138.0,80572126.0


Using apply() will be much faster than iterating manually over rows because pandas is utilizing vectorization.

### Grouped aggregation

When exploring and analyzing data, we often need to perform grouped analysis. For example, we may want to find the mean rating across the various genre of movies in the dataset.

Since many of the movies belong to muultiple genre, there are too many combination of genres. So, we create a new column which keeps only the first genre of each movie:

In [70]:
movies_df['genre_one'] = movies_df['Genre'].str.split(',').str[0]

In [71]:
movies_df.groupby('genre_one')['Rating'].agg(('mean', 'std'))

,mean,std
genre_one,,
Action,6.508662,0.847091
Adventure,6.697059,0.888265
Animation,6.939545,0.822725
Biography,7.103270,0.615411
Comedy,6.579301,0.811387
Crime,6.960520,0.792992
Drama,6.963355,0.815379
Family,6.575000,0.865544
Fantasy,6.073333,0.851891


In [72]:
# we can also obtain the mean of all numeric columns:
(
    movies_df
    .drop('ID', axis = 'columns')
    .groupby('genre_one')
    .mean(numeric_only=True)
)

,Rating,Runtime,Metascore,Votes,Gross
genre_one,,,,,
Action,6.508662,116.259873,54.127302,133757.068790,4.655420e+07
Adventure,6.697059,111.858456,58.425024,105576.055147,3.993996e+07
Animation,6.939545,93.053830,63.501485,119680.146998,6.856429e+07
Biography,7.103270,122.320138,64.098065,89606.306368,2.091005e+07
Comedy,6.579301,104.664965,57.012691,61693.785546,2.284620e+07
Crime,6.960520,112.345297,62.282406,92423.574257,1.343133e+07
Drama,6.963355,114.678308,63.925404,73330.230694,1.452171e+07
Family,6.575000,106.000000,56.937406,67858.500000,5.568126e+06
Fantasy,6.073333,98.616667,52.454073,66293.233333,2.183054e+07


### More `.groupby` examples

In [73]:
ipl_data = {
    'Team': ['Riders', 'Riders', 'Devils', 'Devils', 'Kings',
    'kings', 'Kings', 'Kings', 'Riders', 'Royals', 'Royals', 'Riders'],
    'Rank': [1, 2, 2, 3, 3, 4, 1, 1, 2, 4,1,2],
    'Year': [2014, 2015, 2014, 2015, 2014, 2015, 2016, 2017, 2016, 2014, 2015, 2017],
    'Points':[876, 789, 863, 673, 741, 812, 756, 788, 694, 701, 804, 690]
}
df2 = pd.DataFrame(ipl_data)
df2

,Team,Rank,Year,Points
0,Riders,1,2014,876
1,Riders,2,2015,789
2,Devils,2,2014,863
3,Devils,3,2015,673
4,Kings,3,2014,741
5,kings,4,2015,812
6,Kings,1,2016,756
7,Kings,1,2017,788
8,Riders,2,2016,694
9,Royals,4,2014,701


In [74]:
df2.groupby('Team')    # returns a DataFrameGroupBy object

In [75]:
# returns a dictionary containing group names as keys and list of indices of each group member
df2.groupby('Team').groups

{'Devils': [2, 3], 'Kings': [4, 6, 7], 'Riders': [0, 1, 8, 11], 'Royals': [9, 10], 'kings': [5]}

In [76]:
# can also group by multiple columns
df2.groupby(['Team', 'Year']).groups

{('Devils', 2014): [2], ('Devils', 2015): [3], ('Kings', 2014): [4], ('Kings', 2016): [6], ('Kings', 2017): [7], ('Riders', 2014): [0], ('Riders', 2015): [1], ('Riders', 2016): [8], ('Riders', 2017): [11], ('Royals', 2014): [9], ('Royals', 2015): [10], ('kings', 2015): [5]}

In [77]:
# we can iterate over groups using a loop
grouped = df2.groupby('Year')

for name, group in grouped:
   print(name)
   print(group)

2014
     Team  Rank  Year  Points
0  Riders     1  2014     876
2  Devils     2  2014     863
4   Kings     3  2014     741
9  Royals     4  2014     701
2015
      Team  Rank  Year  Points
1   Riders     2  2015     789
3   Devils     3  2015     673
5    kings     4  2015     812
10  Royals     1  2015     804
2016
     Team  Rank  Year  Points
6   Kings     1  2016     756
8  Riders     2  2016     694
2017
      Team  Rank  Year  Points
7    Kings     1  2017     788
11  Riders     2  2017     690


In [78]:
# we can extract the data of a single group
grouped.get_group(2014)    # a dataframe with Year 2014 only

,Team,Rank,Year,Points
0,Riders,1,2014,876
2,Devils,2,2014,863
4,Kings,3,2014,741
9,Royals,4,2014,701


In [79]:
grouped2 = df2.groupby('Team')
grouped2['Points'].agg('mean')

,Points
Team,
Devils,768.000000
Kings,761.666667
Riders,762.250000
Royals,752.500000
kings,812.000000


In [80]:
grouped2['Points'].agg(['sum', 'mean', 'std'])

,sum,mean,std
Team,,,
Devils,1536,768.000000,134.350288
Kings,2285,761.666667,24.006943
Riders,3049,762.250000,88.567771
Royals,1505,752.500000,72.831998
kings,812,812.000000,NaN


In [81]:
# grouped transformation:
## standardizing the vaiables
grouped2['Points'].transform(lambda x: (x - x.mean()) / x.std()*10)

,Points
0,12.843272
1,3.020286
2,7.071068
3,-7.071068
4,-8.608621
5,NaN
6,-2.360428
7,10.969049
8,-7.705963
9,-7.071068


In [82]:
# filtering in a grouped context
## keeping teams which appears at least three times in the data
grouped2.filter(lambda x: len(x) >= 3)

,Team,Rank,Year,Points
0,Riders,1,2014,876
1,Riders,2,2015,789
4,Kings,3,2014,741
6,Kings,1,2016,756
7,Kings,1,2017,788
8,Riders,2,2016,694
11,Riders,2,2017,690


#### Groupby the summarize/aggregate

In [83]:
data = {
    "state" : ["Ohio"]*3 + ["Nevada"]*3,
    "year" : [2000, 2001, 2002, 2001, 2002, 2003],
    "popn" : [1.5, 1.7, 3.6, 2.4, 2.9, 3.2],
    "wt" : [1, 2, 3] * 2    # weights for findingg weighted average
}
df = pd.DataFrame(data)
df

,state,year,popn,wt
0,Ohio,2000,1.5,1
1,Ohio,2001,1.7,2
2,Ohio,2002,3.6,3
3,Nevada,2001,2.4,1
4,Nevada,2002,2.9,2
5,Nevada,2003,3.2,3


In [84]:
df.groupby("state").agg(avg_pop = ("popn", "mean"))    # returns a dataframe

,avg_pop
state,
Nevada,2.833333
Ohio,2.266667


In [85]:
# finding weighted average
df.groupby("state").apply(lambda g: sum(g.popn * g.wt) / sum(g.wt))    # returns a Series

/tmp/ipython-input-2457851632.py:2: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  df.groupby("state").apply(lambda g: sum(g.popn * g.wt) / sum(g.wt))    # returns a Series


,0
state,
Nevada,2.966667
Ohio,2.616667


In [86]:
# wrapping the result in a Series makes the whole thing a dataframe
df.groupby("state").apply(
    lambda g: pd.Series(
        {"wt_avg" : sum(g.popn * g.wt) / sum(g.wt)}
    )
)

/tmp/ipython-input-1900871211.py:2: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  df.groupby("state").apply(


,wt_avg
state,
Nevada,2.966667
Ohio,2.616667


In [87]:
# this returns a dataframe
agdf = (
    df.groupby(df.state).apply(
        lambda x: pd.Series(
            {
                "avg" : np.mean(x.popn),
                "wt_avg" : sum(x.popn * x.wt)/sum(x.wt),
                "wt_sum" : sum(x.popn * x.wt)
            }
            )
        )
    )

agdf

/tmp/ipython-input-3331310557.py:3: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  df.groupby(df.state).apply(


,avg,wt_avg,wt_sum
state,,,
Nevada,2.833333,2.966667,17.8
Ohio,2.266667,2.616667,15.7


#### Groupby then mutate

In [88]:
df.join(agdf, on = "state")

,state,year,popn,wt,avg,wt_avg,wt_sum
0,Ohio,2000,1.5,1,2.266667,2.616667,15.7
1,Ohio,2001,1.7,2,2.266667,2.616667,15.7
2,Ohio,2002,3.6,3,2.266667,2.616667,15.7
3,Nevada,2001,2.4,1,2.833333,2.966667,17.8
4,Nevada,2002,2.9,2,2.833333,2.966667,17.8
5,Nevada,2003,3.2,3,2.833333,2.966667,17.8


In [89]:
df["avg_pop2"] = df.groupby(df.state)["popn"].transform(lambda x: x.mean())
df

,state,year,popn,wt,avg_pop2
0,Ohio,2000,1.5,1,2.266667
1,Ohio,2001,1.7,2,2.266667
2,Ohio,2002,3.6,3,2.266667
3,Nevada,2001,2.4,1,2.833333
4,Nevada,2002,2.9,2,2.833333
5,Nevada,2003,3.2,3,2.833333


In [90]:
(
    df
    .groupby("state", group_keys = False)
    .apply(
        lambda grp: grp.assign(
            wt_avg2 = lambda x: sum(x.popn * x.wt) / sum(x.wt)
        ),
    )
)

/tmp/ipython-input-268474034.py:4: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  .apply(


,state,year,popn,wt,avg_pop2,wt_avg2
0,Ohio,2000,1.5,1,2.266667,2.616667
1,Ohio,2001,1.7,2,2.266667,2.616667
2,Ohio,2002,3.6,3,2.266667,2.616667
3,Nevada,2001,2.4,1,2.833333,2.966667
4,Nevada,2002,2.9,2,2.833333,2.966667
5,Nevada,2003,3.2,3,2.833333,2.966667
